### PR TITLE
Add `CuPyLogger`

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

Adds a new logger to print out CUDA memory stats pertinent to the `cupy` allocator/memory pool used in Thinc.

### Types of change

New feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -271,6 +272,39 @@ logger2 = {"@loggers": "spacy.LookupLogger.v1", "substring": "pytorch"}
...
+
+- [CUDA memory pool statistics](https://docs.cupy.dev/en/stable/user_guide/memory.html)
+
+### Example config
+
+```ini
+[training.logger]
+@loggers = "spacy.ChainLogger.v1"
+logger1 = {"@loggers": "spacy.CuPyLogger.v1", "prefix": "cupy"}
+logger2 = {"@loggers": "spacy.LookupLogger.v1", "patterns": ["cupy"]}
```
Comment (rmitsch) [inline README.md:301]: The `LookupLogger` is exchangeable and processes the data collected by the `CupyLogger`. Correct?

File: spacy_loggers/cupy.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from typing import Dict, Any, Optional, IO
+import sys
+
+from spacy import Language
+from thinc.backends import context_pools
+from thinc.util import has_cupy_gpu
+from .util import LoggerT
+
+
+def cupy_logger_v1(
```
Comment (rmitsch) [inline spacy_loggers/cupy.py:13]: Please add a docstring here.

**Review by rmitsch — COMMENTED** [review]
Can't speak for the `cupy` details, but LGTM. Will run this locally in a bit before approving.

File: README.md [hunk 3]
```diff
@@ -14,6 +14,7 @@ library.
 - [MLflow](https://www.mlflow.org/)
 - [ClearML](https://www.clear.ml/)
 - [PyTorch](https://pytorch.org/)
+- [CuPy](https://github.com/cupy/cupy)
```
Comment (adrianeboyd) [inline README.md:17]: Since we have `CupyOps`, I'd rather have `CupyLogger`.

File: README.md (hunk 1 above)
Comment (rmitsch) [reply README.md]: Maybe add a comment indicating the the `LookupLogger` can be swapped for any other (equivalent) logger?

Comment (shadeMe) [reply README.md]: Done.

**Review by rmitsch — APPROVED** [review]
LGTM and behaves as expected during training.

File: spacy_loggers/cupy.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        prefix (str):
+            All metric names are prefixed with this string using dot notation, e.g: `<prefix>.<metric>`.
+
+    Returns:
+        LoggerT: Logger instance.
+    """
+    try:
+        import cupy
+
+        if not has_cupy_gpu:
```
Comment (danieldk) [inline spacy_loggers/cupy.py:28]: This should probably emit a different error. The current sets the user on the wrong path, when they did install CuPy, but CUDA could not be initialized. In that case it's also nice to know why.

File: spacy_loggers/cupy.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            # use its own allocator/pool for interop with PyTorch/TensorFlow.
+            cupy_memory_pool = cupy.cuda.get_allocator().__self__
+            thinc_pool = context_pools.get()
+
+            if thinc_pool.get("pytorch") == cupy_memory_pool:
+                pool_source = "pytorch"
+            elif thinc_pool.get("tensorflow") == cupy_memory_pool:
+                pool_source = "tensorflow"
+            else:
+                pool_source = "default"
```
Comment (danieldk) [inline spacy_loggers/cupy.py:55]: It would be nice to have tests for this (conditionally enabled based on the Thinc compat attributes).

File: spacy_loggers/cupy.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+            info[f"{prefix}.pool.source"] = pool_source
+            info[f"{prefix}.pool.acquired_megabytes"] = to_mb(
+                cupy_memory_pool.total_bytes()
+            )
+            info[f"{prefix}.pool.used_megabytes"] = to_mb(cupy_memory_pool.used_bytes())
```
Comment (danieldk) [inline spacy_loggers/cupy.py:61]: Nitpick, but it's probably better to use `mib` here. In SI units megabyte is 10**6 (even though I am very much against it 🙈 ).

File: spacy_loggers/cupy.py (hunk 4 above)
Comment (shadeMe) [reply spacy_loggers/cupy.py:28]: Good point, I'll change that.

**Review by danieldk — COMMENTED** [review]
LGTM. But someone else should check & merge it, since I don't know much about the loggers.

File: README.md [hunk 7]
```diff
@@ -271,6 +272,40 @@ logger2 = {"@loggers": "spacy.LookupLogger.v1", "substring": "pytorch"}
...
+
+This logger requires `cupy` to be installed:
+
+```bash
+pip install cupy
+```
+
+### Usage
+
+Similar to `PyTorchLogger`, `spacy.CupyLogger.v1` is intended to be used with [ChainLogger](#chainlogger).
```
Comment (svlandeg) [inline README.md:287]: Do you think it's helpful to call out the similarity with `PyTorchLogger`? Users might just only reading the `CupyLogger` section, independent from any other information on this page, and feel like they need to look up this explanation from somewhere else.

File: README.md [hunk 8]
```diff
@@ -271,6 +272,40 @@ logger2 = {"@loggers": "spacy.LookupLogger.v1", "substring": "pytorch"}
...
+[training.logger]
+@loggers = "spacy.ChainLogger.v1"
+logger1 = {"@loggers": "spacy.CupyLogger.v1", "prefix": "cupy"}
+# Alternatively, you can use any other logger that provides the `log_custom_stats` parameter.
+logger2 = {"@loggers": "spacy.LookupLogger.v1", "patterns": ["cupy"]}
+```
+
+| Name     | Type  | Description                                                                                                    |
+| -------- | ----- | -------------------------------------------------------------------------------------------------------------- |
+| `prefix` | `str` | All metric names are prefixed with this string using dot notation, e.g: `<prefix>.<metric>` (default: `cupy`). |
```
Comment (svlandeg) [inline README.md:307]:
```suggestion
| `prefix` | `str` | All metric names are prefixed with this string using dot notation, e.g: `<prefix>.<metric>` (default: `"cupy"`). |
```

File: spacy_loggers/cupy.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        prefix (str):
+            All metric names are prefixed with this string using dot notation, e.g: `<prefix>.<metric>`.
+
+    Returns:
+        LoggerT: Logger instance.
+    """
+    try:
+        import cupy
+
+        if not has_cupy_gpu:
```
Comment (svlandeg) [inline spacy_loggers/cupy.py:28]: Why isn't the `has_cupy_gpu` check outside the `try` block? It's not like this will throw an `ImportError`, right?

**Review by svlandeg — COMMENTED** [review]
Looks good, just had a few small questions/suggestions.

**Review by danieldk — APPROVED** [review]
